calypso/contracts.go

```diff
@@ -0,0 +1,127 @@
+package calypso
+
+import (
+	"errors"
+
+	"github.com/dedis/cothority"
+	"github.com/dedis/cothority/omniledger/darc"
+	ol "github.com/dedis/cothority/omniledger/service"
+	"github.com/dedis/onet/log"
+	"github.com/dedis/onet/network"
+	"github.com/dedis/protobuf"
+)
+
+// ContractWriteID references a write contract system-wide.
+var ContractWriteID = "calypsoWrite"
+
+// ContractWrite is used to store a secret in OmniLedger, so that an
+// authorized reader can retrieve it by creating a Read-instance.
+//
+// Accepted Instructions:
+//  - spawn:calypsoWrite creates a new write-request. TODO: verify the LTS exists
+//  - spawn:calypsoRead creates a new read-request for this write-request.
+func (s *Service) ContractWrite(cdb ol.CollectionView, inst ol.Instruction, c []ol.Coin) ([]ol.StateChange, []ol.Coin, error) {
+	err := inst.VerifyDarcSignature(cdb)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var darcID darc.ID
+	_, _, darcID, err = cdb.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	switch inst.GetType() {
+	case ol.SpawnType:
+		var sc ol.StateChanges
+		nc := c
+		switch inst.Spawn.ContractID {
+		case ContractWriteID:
+			w := inst.Spawn.Args.Search("write")
+			if w == nil || len(w) == 0 {
+				return nil, nil, errors.New("need a write request in 'write' argument")
+			}
+			var wr Write
+			err := protobuf.DecodeWithConstructors(w, &wr, network.DefaultConstructors(cothority.Suite))
+			if err != nil {
+				return nil, nil, errors.New("couldn't unmarshal write: " + err.Error())
+			}
+			if err = wr.CheckProof(cothority.Suite, darcID); err != nil {
+				return nil, nil, errors.New("proof of write failed: " + err.Error())
+			}
+			instID := inst.DeriveID("")
+			log.Lvlf3("Successfully verified write request and will store in %x", instID)
+			sc = append(sc, ol.NewStateChange(ol.Create, instID, ContractWriteID, w, darcID))
+		case ContractReadID:
```

Comment:
This is what was here, I believe you have to send the read request to the write request in order to spawn read. I think this is what's in the test and also in the diagram. Which might be wrong, because the reader might not have the same authorisation in the darc as the writer, can you confirm please?

Reply:
Now I see the light. This is correct. And using a rule like Spawn:ContractReadID in the darc can give access rights as needed.